homeassistant/components/history_stats/sensor.py

```diff
@@ -96,214 +96,72 @@ async def async_setup_platform(
-                # If period has not changed and current time after the period end...
-                # Don't compute anything as the value cannot have changed
-                return
-        else:
-            self._history_current_period = await get_instance(
-                self.hass
-            ).async_add_executor_job(
-                self._update_from_database,
-                current_period_start,
-                current_period_end,
-            )
-            self._previous_run_before_start = False
-
-        if not self._history_current_period:
-            self._async_set_native_value(None, None)
-            return
-
-        hours_matched, changes_to_match_state = self._async_compute_hours_and_changes(
-            now_timestamp,
-            current_period_start_timestamp,
-            current_period_end_timestamp,
-        )
-        self._async_set_native_value(hours_matched, changes_to_match_state)
+    @abstractmethod
+    def _process_update(self) -> None:
+        """Process an update from the coordinator."""
 
-    def _update_from_database(
-        self, start: datetime.datetime, end: datetime.datetime
-    ) -> list[State]:
-        return history.state_changes_during_period(
-            self.hass,
-            start,
-            end,
-            self._entity_id,
-            include_start_time_state=True,
-            no_attributes=True,
-        ).get(self._entity_id, [])
 
-    def _async_compute_hours_and_changes(
-        self, now_timestamp: float, start_timestamp: float, end_timestamp: float
-    ) -> tuple[float, int]:
-        """Compute the hours matched and changes from the history list and first state."""
-        # state_changes_during_period is called with include_start_time_state=True
-        # which is the default and always provides the state at the start
-        # of the period
-        previous_state_matches = (
-            self._history_current_period
-            and self._history_current_period[0].state in self._entity_states
-        )
-        last_state_change_timestamp = start_timestamp
-        elapsed = 0.0
-        changes_to_match_state = 0
-
-        # Make calculations
-        for item in self._history_current_period:
-            current_state_matches = item.state in self._entity_states
-            state_change_timestamp = item.last_changed.timestamp()
-
-            if previous_state_matches:
-                elapsed += state_change_timestamp - last_state_change_timestamp
-            elif current_state_matches:
-                changes_to_match_state += 1
-
-            previous_state_matches = current_state_matches
-            last_state_change_timestamp = state_change_timestamp
-
-        # Count time elapsed between last history state and end of measure
-        if previous_state_matches:
-            measure_end = min(end_timestamp, now_timestamp)
-            elapsed += measure_end - last_state_change_timestamp
+class HistoryStatsSensor(HistoryStatsSensorBase):
```

Comment:
Should we set a state class so we generate stats ?

Reply:
added to the list of the larger tracking PR https://github.com/home-assistant/core/pull/70131

Reply:
https://github.com/home-assistant/core/pull/70720